org.eclipse.jdt.core/codeassist/org/eclipse/jdt/internal/codeassist/CompletionEngine.java

```diff
@@ -3365,6 +3356,30 @@ private void findCompletionsForArgumentPosition(ObjectVector methodsFound, int c
 		}
 	}
 
+	private void pushExpectedTypesForArgumentPosition(ObjectVector methodsToSearchOn, int completedArgumentLength,
+			Scope scope) {
+		if (methodsToSearchOn.size == 0) {
+			return;
+		}
+		for (int i = 0; i < methodsToSearchOn.size; i++) {
+			MethodBinding method = (MethodBinding) ((Object[]) methodsToSearchOn.elementAt(i))[0];
+			if (method.parameters.length <= completedArgumentLength && !method.isVarargs()) {
+				continue;
+			}
+
+			int index = (method.parameters.length <= completedArgumentLength)
+					? method.parameters.length - 1
+					: completedArgumentLength;
+
+			TypeBinding paramType = method.parameters[index];
+			addExpectedType(paramType, scope);
+			if (method.isVarargs() && paramType.isArrayType()
+					&& index == (method.parameters.length - 1)) {
```

Comment:
IMO `(method.parameters.length <= completedArgumentLength)` should be extracted to some local variable, I see it used in 2 places above and (indirectly) here.